File: DSA/Hima-Swarupa-Padala/day9q1.py
class MaxStack:
    def __init__(self):
        self.stack = [] 
        self.max_stack = []

    def push(self, val):
        self.stack.append(val)
        if not self.max_stack or val >= self.max_stack[-1]:
            self.max_stack.append(val)

    def pop(self):
        if self.stack:
            popped = self.stack.pop()
            if popped == self.max_stack[-1]:
                self.max_stack.pop()
            return popped
        else:
            return None

    def get_max(self):
        return self.max_stack[-1] if self.max_stack else None

stack = MaxStack()
stack.push(1)
stack.push(2)
print(stack.get_max())
stack.pop()
print(stack.get_max())

File: DSA/Shruti-Pandey/Day_9_q1.java
import java.util.Stack;
class MaxStack {
    private Stack<Integer> mainStack;
    private Stack<Integer> maxStack;

    public MaxStack() {
        mainStack = new Stack<>();
        maxStack = new Stack<>();
    }

    public void push(int x) {
        mainStack.push(x);
        if (maxStack.isEmpty() || x >= maxStack.peek()) {
            maxStack.push(x);
        }
    }

    public int pop() {
        if (mainStack.isEmpty()) {
            throw new IllegalStateException("Stack is empty");
        }
        int popped = mainStack.pop();
        if (popped == maxStack.peek()) {
            maxStack.pop();
        }
        return popped;
    }

    public int getMax() {
        if (maxStack.isEmpty()) {
            throw new IllegalStateException("Stack is empty");
        }
        return maxStack.peek();
    }

    public boolean isEmpty() {
        return mainStack.isEmpty();
    }
}

public class Main {
    public static void main(String[] args) {
        MaxStack stack = new MaxStack();
        
        stack.push(1);
        stack.push(2);
        System.out.println("Max: " + stack.getMax()); 
        stack.pop();
        System.out.println("Max: " + stack.getMax()); 
    }
}
